File: ruby/thread/thread_sample.rb
#!/usr/bin/ruby
# -*- coding: utf-8 -*-

require "rubygems"
require "zipruby"
require "thread"

# url
## http://www.namaraii.com/rubytips/?%E3%82%B9%E3%83%AC%E3%83%83%E3%83%89
## http://techracho.bpsinc.jp/piichan1031/2010_07_05/2030

# unzip function
def unzip(path)
  Zip::Archive.open(path) do |archives|
    archives.each do |a|
      unless a.directory?
        # reading only
        a.read
      end
    end
  end
end

# queue
queue = Queue.new

# push task 
path = "/home/kzk/downloads/*.zip"
Dir.glob(path).each do |p|
  queue.push(p)
end

# create/start therad
num_threads = 3
st = Time.now
(0..(num_threads - 1)).each do |e|
  t = Thread.new() do
    loop do
      p = queue.pop
      unzip(p)
    end
  end
end

# queue empty
while true
  if queue.empty?
    break
  end
end
et = Time.now

puts "total execution time with threading: #{et - st} [s]"

# total execution time with threading: 51.790661849 [s]

File: python/thread/thread_sample.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

# url
## http://ja.pymotw.com/2/threading/
## http://docs.python.jp/2.7/library/threading.html

import threading
import glob
import time
import zipfile
import Queue

# http://docs.python.jp/2/library/queue.html#module-Queue

# read onley unzip
def unzip(zfin):
    zf = zipfile.ZipFile(zfin, "r")
    for f in zf.namelist():  # zipは複数ファイルが１つにまとめられている前提のため
        #print "unzip", f
        zf.read(f)  # reading only
    zf.close
    pass

# queue
## targetにinstanceを渡せない
## 渡したいなら，Threadクラスを継承したクラスを作る
queue = Queue.Queue()

# wokrer
def worker():
    """
    """
    while True:
        path = queue.get()
        unzip(path)
        queue.task_done()
        pass

    pass

# put task
path = "/home/kzk/downloads/*.zip"
for path in glob.glob(path):
    queue.put(path)
    pass

# create/start  thread
st = time.time()
num_worker_threads = 3
for i in range(num_worker_threads):
    t = threading.Thread(target=worker)
    t.daemon = True
    t.start()
    pass

queue.join()  # wait
et = time.time()
print "total execution time with threading: ", (et - st), "[s]"

# total execution time with threading:  60.7147810459 [s]

